test/device_selection_tests.cc

```diff
@@ -0,0 +1,641 @@
+		tp_0.set_devices({td_1});
+		tp_0.set_info("foo_0");
+
+		mock_platform tp_1(1);
+		mock_device td_2(0, tp_1, cl::sycl::info::device_type::gpu);
+		tp_1.set_devices({td_2});
+		tp_1.set_info("foo_1");
+
+		mock_platform tp_2(1);
+		mock_device td_5(0, tp_2, cl::sycl::info::device_type::accelerator);
+		tp_2.set_devices({td_5});
+		tp_2.set_info("foo_2");
+
+		size_t node_count = 4;
+		size_t local_rank = 3;
+		size_t local_num_cpus = 1;
+
+		celerity::detail::host_config h_cfg{node_count, local_rank, local_num_cpus};
+		celerity::detail::config_testspy::set_mock_host_cfg(cfg, h_cfg);
+		auto device_selector = [td_5](const mock_device& d) -> int { return d == td_5 ? 2 : 1; };
+
+		auto device = pick_device(cfg, device_selector, std::vector<mock_platform>{tp_0, tp_1, tp_2});
+		std::vector<mock_device> devices = tp_2.get_devices();
+		CHECK_THAT(lc.get_log(),
+		    Catch::Matchers::ContainsSubstring("No suitable platform found that can provide 4 devices that matches the specified device selector"));
+		CHECK(device == devices[0]);
+	}
+
+	SECTION("pick_device doesn't choose devices across platform") {
+		log_capture lc;
+
+		mock_platform tp_0(0);
+		mock_device td_1(0, tp_0, cl::sycl::info::device_type::cpu);
+		tp_0.set_devices({td_1});
+		tp_0.set_info("foo_0");
+
+		mock_platform tp_1(1);
+		mock_device td_2(0, tp_1, cl::sycl::info::device_type::accelerator);
+		mock_device td_3(1, tp_1, cl::sycl::info::device_type::gpu);
+		mock_device td_4(2, tp_1, cl::sycl::info::device_type::gpu);
+		mock_device td_5(3, tp_1, cl::sycl::info::device_type::gpu);
+		tp_1.set_devices({td_2, td_3, td_4});
+		tp_1.set_info("foo_1");
+
+		mock_platform tp_2(2);
+		mock_device td_8(2, tp_2, cl::sycl::info::device_type::gpu);
+		mock_device td_9(3, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_10(4, tp_2, cl::sycl::info::device_type::accelerator);
+		mock_device td_11(5, tp_2, cl::sycl::info::device_type::accelerator);
+		tp_2.set_devices({td_8, td_9, td_10, td_11});
+		tp_2.set_info("foo_2");
+
+		size_t node_count = 4;
+		size_t local_rank = 2;
+		size_t local_num_cpus = 1;
+
+		celerity::detail::host_config h_cfg{node_count, local_rank, local_num_cpus};
+		celerity::detail::config_testspy::set_mock_host_cfg(cfg, h_cfg);
+
+		auto device_selector = [](const mock_device& d) -> int { return d.get_type() == cl::sycl::info::device_type::accelerator ? 2 : 1; };
+
+		auto device = pick_device(cfg, device_selector, std::vector<mock_platform>{tp_0, tp_1, tp_2});
+		std::vector<mock_device> devices = tp_2.get_devices();
+		INFO("Platform id" << device.get_platform().get_id() << " device id " << device.get_id());
+		CHECK_THAT(lc.get_log(), Catch::Matchers::ContainsSubstring("device selector specified: platform"));
+		CHECK(device == devices[3]);
+	}
+}
+
+TEST_CASE_METHOD(celerity::test_utils::mpi_fixture, "pick_device fails on returning a negative number for a device with device selector",
```

Comment:
The test name is pretty unclear IMO, I'd suggest `"pick_device does not consider devices with a negative selector score"` or something.